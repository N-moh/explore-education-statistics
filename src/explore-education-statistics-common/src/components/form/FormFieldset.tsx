import classNames from 'classnames';
import React, { ReactNode } from 'react';
import ErrorMessage from '../ErrorMessage';
import FormGroup from './FormGroup';
import createDescribedBy from './util/createDescribedBy';

export interface FieldSetProps {
  children?: ReactNode;
  error?: string;
  hint?: string;
  id: string;
  legend?: ReactNode | string;
  legendSize?: 'xl' | 'l' | 'm' | 's';
}

const FormFieldset = ({
  children,
  error,
  hint,
  id,
  legend,
  legendSize = 'm',
}: FieldSetProps) => {
  return (
    <FormGroup hasError={!!error}>
      <fieldset
        className="govuk-fieldset"
        id={id}
        aria-describedby={createDescribedBy({
          id,
          error: !!error,
          hint: !!hint,
        })}
      >
        {legend && (
          <legend
            className={classNames(
              'govuk-fieldset__legend',
              `govuk-fieldset__legend--${legendSize}`,
            )}
          >
            {legend}
          </legend>
        )}

        {hint && (
          <span className="govuk-hint" id={`${id}-hint`}>
            {hint}
          </span>
        )}

        {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}

        {children}
      </fieldset>
    </FormGroup>
  );
};

export default FormFieldset;
