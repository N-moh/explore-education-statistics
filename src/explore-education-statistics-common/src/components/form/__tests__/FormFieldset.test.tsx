import React from 'react';
import { render } from 'react-testing-library';
import FormFieldset from '../FormFieldset';

describe('FormFieldset', () => {
  test('renders correctly with required props', () => {
    const { container } = render(
      <FormFieldset id="test-fieldset" legend="Fill the form">
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    const legend = container.querySelector('legend') as HTMLLegendElement;

    expect(legend.textContent).toBe('Fill the form');
    expect(container.querySelector('#test-fieldset')).not.toBeNull();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with hint', () => {
    const { container, getByText } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    expect(getByText('All fields are required')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with error', () => {
    const { container, getByText } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        error="There was an error"
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    expect(getByText('There was an error')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('aria-describedby is equal to the hint id', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
      />,
    );

    expect(container.querySelector('fieldset')).toHaveAttribute(
      'aria-describedby',
      'test-fieldset-hint',
    );
  });

  test('aria-describedby is equal to the error id', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        error="There was an error"
      />,
    );

    expect(container.querySelector('fieldset')).toHaveAttribute(
      'aria-describedby',
      'test-fieldset-error',
    );
  });

  test('aria-describedby contains both hint and error ids', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
        error="There was an error"
      />,
    );

    const fieldset = container.querySelector('fieldset');
    const ariaDescribedBy = fieldset
      ? fieldset.getAttribute('aria-describedby')
      : '';

    expect(ariaDescribedBy).toContain('test-fieldset-error');
    expect(ariaDescribedBy).toContain('test-fieldset-hint');
  });

  test('setting `legendSize` to xl applies class correctly', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
        legendSize="xl"
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    expect(container.querySelector('legend')).toHaveClass(
      'govuk-fieldset__legend--xl',
    );
  });

  test('setting `legendHidden` to true hides the legend', () => {
    const { container } = render(
      <FormFieldset
        id="test-fieldset"
        legend="Fill the form"
        hint="All fields are required"
        legendHidden
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldset>,
    );

    expect(container.querySelector('legend')).toHaveClass(
      'govuk-visually-hidden',
    );
  });
});
