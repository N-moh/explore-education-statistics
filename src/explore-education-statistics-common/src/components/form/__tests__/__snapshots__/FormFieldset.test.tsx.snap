// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldset renders correctly with error 1`] = `
<div class="govuk-form-group govuk-form-group--error">
  <fieldset class="govuk-fieldset"
            id="test-fieldset"
            aria-describedby="test-fieldset-error"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Fill the form
    </legend>
    <span class="govuk-error-message"
          id="test-fieldset-error"
    >
      There was an error
    </span>
    <input type="text"
           name="test-input-1"
    >
    <input type="text"
           name="test-input-2"
    >
  </fieldset>
</div>
`;

exports[`FormFieldset renders correctly with hint 1`] = `
<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-fieldset"
            aria-describedby="test-fieldset-hint"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Fill the form
    </legend>
    <span class="govuk-hint"
          id="test-fieldset-hint"
    >
      All fields are required
    </span>
    <input type="text"
           name="test-input-1"
    >
    <input type="text"
           name="test-input-2"
    >
  </fieldset>
</div>
`;

exports[`FormFieldset renders correctly with required props 1`] = `
<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-fieldset"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Fill the form
    </legend>
    <input type="text"
           name="test-input-1"
    >
    <input type="text"
           name="test-input-2"
    >
  </fieldset>
</div>
`;
